### ISSUE TITLE
Can Custom Providers Override GPS Sensors?

### ISSUE THREAD

**Issue by taylorbrown2686** [dmbfm/unity-ar-gps-location-issues#81]
My name is Taylor and I am from a startup called Earth Media. We are currently creating AR experiences outside, and we are working on accurate GPS to position our objects. Here is our idea:

Create a custom provider similar to the 'mock' one. The player will manually set their coordinate depending on where they are located, and it will be handled by the accelerometer and gyroscope after that point. We believe this is the solution to what we have been trying, but we can't seem to find out how to manually set our position at a GPS we define instead of relying on the sensor data.

Is there any way with your plugin to ignore the sensor data and use our own values to calculate GPS position and updated position? Please let me know as soon as possible, thank you for your time!

**Comment by dmbfm** [comment]
Yes. This section of the documentation describes how to implement a custom location provider: https://docs.unity-ar-gps-location.com/guide/#custom-location-providers. It seems
this is what you guys need. Let me  if I can help you further on this,

Thanks!

**Comment by taylorbrown2686** [comment]
Yes, we have seen that. It isn't very helpful since the default Unity provider inherits from an Abstract provider, which inherits from the base class. It appears that both are required to make a working provider, but ours can also be much simpler as we simply want to spoof the location. I assume we can just inherit from ILocationProvider in this case, but there still leaves the question: How do I spoof my location (at a coordinate)? In other words, how do I ignore Input.location.lastData and the GPS sensor on the phone, and simply input a coordinate that the engine then recognizes as my position?

Thank you for the help,
Taylor Brown

**Comment by dmbfm** [comment]
To create a Location Provider you just need a class that implements `ILocationProvider`...  as it says on the documentation, you can take a look at the `MockLocationProvider` source code as an example of how to do that.

**Comment by taylorbrown2686** [comment]
We have also already attempted that. When we do, the console is spammed with 'Location updates are set to off, please turn them on before querying location'. This is the issue we are trying to resolve, we don't care that location updates are off, we just want to manually set the coordinates and update them via math based on gyro/accelerometer sensors.

**Comment by dmbfm** [comment]
Can you show me the code you used for that?

**Comment by taylorbrown2686** [comment]
`﻿using System.Collections;
using System.Collections.Generic;
using UnityEngine;
//This is a custom location provider for AR + GPS. It allows us to set a position manually
namespace ARLocation {

  public class AIVRACustomProvider : AbstractLocationProvider
  {
      //This provider should start with no coordinate being reported. Once a GPS position is touched, a public method
      //to set the position to the one touched will be called, and it will set the GPS position to that
      //Sensor data should be ignored until the GPS is set. After that, accel/gyro will handle the rest.
      public override string Name => "AIVRACustomProvider";

      public override bool IsCompassEnabled => true;

      public Location mockLocation = new Location();

      private void Start() {
        mockLocation.Latitude = 0;
        mockLocation.Longitude = 0;
        mockLocation.Altitude = 0;
        RequestLocationAndCompassUpdates();
      }

      protected override HeadingReading? ReadHeading()
      {
          var mainCamera = ARLocationManager.Instance.MainCamera;

          var transform = mainCamera.transform;

          var localEulerAngles = transform.localEulerAngles;
          return new HeadingReading
          {
              heading = localEulerAngles.y,
              magneticHeading = localEulerAngles.y,
              accuracy = 0,
              isMagneticHeadingAvailable = true,
              timestamp = (long)(Time.time * 1000)
          };
      }

      protected override LocationReading? ReadLocation()
      {
          return new LocationReading
          {
              latitude = mockLocation.Latitude,
              longitude = mockLocation.Longitude,
              altitude = mockLocation.Altitude,
              accuracy = 0.0,
              floor = -1,
              timestamp = (long)(Time.time * 1000)
          };
      }

      private bool requested = true;

      protected override void RequestLocationAndCompassUpdates()
      {
          requested = true;
      }

      protected override void UpdateLocationRequestStatus()
      {
          if (requested)
          {
              Status = LocationProviderStatus.Initializing;
              requested = false;
          }

          if (Status == LocationProviderStatus.Initializing)
          {
              Status = LocationProviderStatus.Started;
          }
      }

      public void SetSpoofedLocation(float lat, float lng) {
        mockLocation.Latitude = lat;
        mockLocation.Longitude = lng;
        mockLocation.Altitude = 0;
      }
  }

}
`

Although it's been changed a lot with different methods being tried, this is the current version.

**Comment by dmbfm** [comment]
If you just add the `Input.location.Start()` at the start method, does this goes away? This log message does not come from this asset. I'm thinking maybe this happens because the magnetic compass data is being requested without location services being enabled, causing this warning message.
